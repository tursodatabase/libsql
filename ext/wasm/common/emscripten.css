/* emscripten-related styling, used during the module load/intialization processes... */
.emscripten { padding-right: 0; margin-left: auto; margin-right: auto; display: block; }
div.emscripten { text-align: center; }
div.emscripten_border { border: 1px solid black; }
#module-spinner { overflow: visible; }
#module-spinner > * {
    margin-top: 1em;
}
.spinner {
    height: 50px;
    width: 50px;
    margin: 0px auto;
    animation: rotation 0.8s linear infinite;
    border-left: 10px solid rgb(0,150,240);
    border-right: 10px solid rgb(0,150,240);
    border-bottom: 10px solid rgb(0,150,240);
    border-top: 10px solid rgb(100,0,200);
    border-radius: 100%;
    background-color: rgb(200,100,250);
}
@keyframes rotation {
    from {transform: rotate(0deg);}
    to {transform: rotate(360deg);}
}
