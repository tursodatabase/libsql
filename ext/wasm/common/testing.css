body {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
}
textarea {
    font-family: monospace;
}
header {
    font-size: 130%;
    font-weight: bold;
}
.hidden, .initially-hidden {
    position: absolute !important;
    opacity: 0 !important;
    pointer-events: none !important;
    display: none !important;
}
fieldset.options {
    font-size: 75%;
}
fieldset > legend {
    padding: 0 0.5em;
}
span.labeled-input {
    padding: 0.25em;
    margin: 0.25em 0.5em;
    border-radius: 0.25em;
    white-space: nowrap;
    background: #0002;
}
.center { text-align: center; }
.error {
    color: red;
    background-color: yellow;
}
.strong { font-weight: 700 }
.warning { color: firebrick; }
.green { color: darkgreen; }
.tests-pass { background-color: green; color: white }
.tests-fail { background-color: red; color: yellow }
.faded { opacity: 0.5; }
.group-start { color: blue; }
.group-end { color: blue; }
.input-wrapper {
  white-space: nowrap;
  display: flex;
  align-items: center;
}
#test-output {
  border: 1px inset;
  border-radius: 0.25em;
  padding: 0.25em;
  /*max-height: 30em;*/
  overflow: auto;
  white-space: break-spaces;
  display: flex; flex-direction: column;
  font-family: monospace;
}
#test-output.reverse {
  flex-direction: column-reverse;
}
label[for] { cursor: pointer }
