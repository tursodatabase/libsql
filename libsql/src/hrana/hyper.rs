use crate::hrana::connection::HttpConnection;
use crate::hrana::pipeline::ServerMsg;
use crate::hrana::proto::Stmt;
use crate::hrana::{HranaError, HttpSend, Result};
use crate::params::Params;
use crate::util::ConnectorService;
use crate::{Rows, Statement};
use futures::future::BoxFuture;
use http::header::AUTHORIZATION;
use http::{HeaderValue, StatusCode};
use hyper_rustls::{HttpsConnector, HttpsConnectorBuilder};

#[derive(Clone, Debug)]
pub struct HttpSender {
    inner: hyper::Client<HttpsConnector<ConnectorService>, hyper::Body>,
    version: HeaderValue,
}

impl HttpSender {
    pub fn new(connector: ConnectorService, version: Option<&str>) -> Self {
        let ver = version.unwrap_or_else(|| env!("CARGO_PKG_VERSION"));

        let version = HeaderValue::try_from(format!("libsql-remote-{ver}")).unwrap();

        let https = HttpsConnectorBuilder::new()
            .with_native_roots()
            .https_or_http()
            .enable_http1()
            .wrap_connector(connector);
        let inner = hyper::Client::builder().build(https);

        Self { inner, version }
    }

    async fn send(&self, url: String, auth: String, body: String) -> Result<ServerMsg> {
        let req = hyper::Request::post(url)
            .header(AUTHORIZATION, auth)
            .header("x-libsql-client-version", self.version.clone())
            .body(hyper::Body::from(body))
            .unwrap();

        let res = self.inner.request(req).await.map_err(HranaError::from)?;

        if res.status() != StatusCode::OK {
            let body = hyper::body::to_bytes(res.into_body())
                .await
                .map_err(HranaError::from)?;
            let body = String::from_utf8(body.into()).unwrap();
            return Err(HranaError::Api(body));
        }

        let body = hyper::body::to_bytes(res.into_body())
            .await
            .map_err(HranaError::from)?;

        let msg = serde_json::from_slice::<ServerMsg>(&body[..]).map_err(HranaError::from)?;

        Ok(msg)
    }
}

impl<'a> HttpSend<'a> for HttpSender {
    type Result = BoxFuture<'a, Result<ServerMsg>>;

    fn http_send(&'a self, url: String, auth: String, body: String) -> Self::Result {
        let fut = self.send(url, auth, body);
        Box::pin(fut)
    }
}

impl From<hyper::Error> for HranaError {
    fn from(value: hyper::Error) -> Self {
        HranaError::Http(value.to_string())
    }
}

impl HttpConnection<HttpSender> {
    pub(crate) fn new_with_connector(
        url: impl Into<String>,
        token: impl Into<String>,
        connector: ConnectorService,
        version: Option<&str>,
    ) -> Self {
        let inner = HttpSender::new(connector, version);
        Self::new(url.into(), token.into(), inner)
    }
}

#[async_trait::async_trait]
impl crate::connection::Conn for HttpConnection<HttpSender> {
    async fn execute(&self, sql: &str, params: Params) -> crate::Result<u64> {
        let mut stmt = self.prepare(sql);
        let rows = stmt.execute(&params).await?;

        Ok(rows as u64)
    }

    async fn execute_batch(&self, sql: &str) -> crate::Result<()> {
        let mut statements = Vec::new();
        let stmts = crate::parser::Statement::parse(sql);
        for s in stmts {
            let s = s?;
            statements.push(Stmt::new(s.stmt, false));
        }
        self.raw_batch(statements)
            .await
            .map_err(|e| crate::Error::Hrana(e.into()))?;
        Ok(())
    }

    async fn prepare(&self, sql: &str) -> crate::Result<Statement> {
        let stmt = crate::hrana::Statement::new(self.clone(), sql.to_string(), true);
        Ok(Statement {
            inner: Box::new(stmt),
        })
    }

    async fn transaction(
        &self,
        _tx_behavior: crate::TransactionBehavior,
    ) -> crate::Result<crate::transaction::Transaction> {
        todo!()
    }

    fn is_autocommit(&self) -> bool {
        // TODO: Is this correct?
        false
    }

    fn changes(&self) -> u64 {
        self.affected_row_count()
    }

    fn last_insert_rowid(&self) -> i64 {
        self.last_insert_rowid()
    }

    fn close(&mut self) {
        todo!()
    }
}

#[async_trait::async_trait]
impl crate::statement::Stmt for crate::hrana::Statement<HttpSender> {
    fn finalize(&mut self) {}

    async fn execute(&mut self, params: &Params) -> crate::Result<usize> {
        self.execute(params).await
    }

    async fn query(&mut self, params: &Params) -> crate::Result<Rows> {
        self.query(params).await
    }

    fn reset(&mut self) {}

    fn parameter_count(&self) -> usize {
        todo!()
    }

    fn parameter_name(&self, _idx: i32) -> Option<&str> {
        todo!()
    }

    fn columns(&self) -> Vec<crate::Column> {
        todo!()
    }
}
