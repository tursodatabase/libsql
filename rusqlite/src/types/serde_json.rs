//! [`ToSql`] and [`FromSql`] implementation for JSON `Value`.

use serde_json::Value;

use crate::types::{FromSql, FromSqlError, FromSqlResult, ToSql, ToSqlOutput, ValueRef};
use crate::Result;

/// Serialize JSON `Value` to text.
impl ToSql for Value {
    #[inline]
    fn to_sql(&self) -> Result<ToSqlOutput<'_>> {
        Ok(ToSqlOutput::from(serde_json::to_string(self).unwrap()))
    }
}

/// Deserialize text/blob to JSON `Value`.
impl FromSql for Value {
    #[inline]
    fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self> {
        match value {
            ValueRef::Text(s) => serde_json::from_slice(s),
            ValueRef::Blob(b) => serde_json::from_slice(b),
            _ => return Err(FromSqlError::InvalidType),
        }
        .map_err(|err| FromSqlError::Other(Box::new(err)))
    }
}

#[cfg(test)]
mod test {
    use crate::types::ToSql;
    use crate::{Connection, Result};

    fn checked_memory_handle() -> Result<Connection> {
        let db = Connection::open_in_memory()?;
        db.execute_batch("CREATE TABLE foo (t TEXT, b BLOB)")?;
        Ok(db)
    }

    #[test]
    fn test_json_value() -> Result<()> {
        let db = checked_memory_handle()?;

        let json = r#"{"foo": 13, "bar": "baz"}"#;
        let data: serde_json::Value = serde_json::from_str(json).unwrap();
        db.execute(
            "INSERT INTO foo (t, b) VALUES (?, ?)",
            &[&data as &dyn ToSql, &json.as_bytes()],
        )?;

        let t: serde_json::Value = db.query_row("SELECT t FROM foo", [], |r| r.get(0))?;
        assert_eq!(data, t);
        let b: serde_json::Value = db.query_row("SELECT b FROM foo", [], |r| r.get(0))?;
        assert_eq!(data, b);
        Ok(())
    }
}
