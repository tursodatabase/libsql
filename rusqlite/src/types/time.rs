//! [`ToSql`] and [`FromSql`] implementation for [`time::OffsetDateTime`].
use crate::types::{FromSql, FromSqlError, FromSqlResult, ToSql, ToSqlOutput, ValueRef};
use crate::Result;
use time::{OffsetDateTime, PrimitiveDateTime, UtcOffset};

const CURRENT_TIMESTAMP_FMT: &str = "%Y-%m-%d %H:%M:%S";
const SQLITE_DATETIME_FMT: &str = "%Y-%m-%dT%H:%M:%S.%NZ";
const SQLITE_DATETIME_FMT_LEGACY: &str = "%Y-%m-%d %H:%M:%S:%N %z";

impl ToSql for OffsetDateTime {
    #[inline]
    fn to_sql(&self) -> Result<ToSqlOutput<'_>> {
        let time_string = self.to_offset(UtcOffset::UTC).format(SQLITE_DATETIME_FMT);
        Ok(ToSqlOutput::from(time_string))
    }
}

impl FromSql for OffsetDateTime {
    fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self> {
        value.as_str().and_then(|s| {
            match s.len() {
                19 => PrimitiveDateTime::parse(s, CURRENT_TIMESTAMP_FMT).map(|d| d.assume_utc()),
                _ => PrimitiveDateTime::parse(s, SQLITE_DATETIME_FMT)
                    .map(|d| d.assume_utc())
                    .or_else(|err| {
                        OffsetDateTime::parse(s, SQLITE_DATETIME_FMT_LEGACY).map_err(|_| err)
                    }),
            }
            .map_err(|err| FromSqlError::Other(Box::new(err)))
        })
    }
}

#[cfg(test)]
mod test {
    use crate::{Connection, Result};
    use std::time::Duration;
    use time::OffsetDateTime;

    fn checked_memory_handle() -> Result<Connection> {
        let db = Connection::open_in_memory()?;
        db.execute_batch("CREATE TABLE foo (t TEXT, i INTEGER, f FLOAT)")?;
        Ok(db)
    }

    #[test]
    fn test_offset_date_time() -> Result<()> {
        let db = checked_memory_handle()?;

        let mut ts_vec = vec![];

        let make_datetime =
            |secs, nanos| OffsetDateTime::from_unix_timestamp(secs) + Duration::from_nanos(nanos);

        ts_vec.push(make_datetime(10_000, 0)); //January 1, 1970 2:46:40 AM
        ts_vec.push(make_datetime(10_000, 1000)); //January 1, 1970 2:46:40 AM (and one microsecond)
        ts_vec.push(make_datetime(1_500_391_124, 1_000_000)); //July 18, 2017
        ts_vec.push(make_datetime(2_000_000_000, 2_000_000)); //May 18, 2033
        ts_vec.push(make_datetime(3_000_000_000, 999_999_999)); //January 24, 2065
        ts_vec.push(make_datetime(10_000_000_000, 0)); //November 20, 2286

        for ts in ts_vec {
            db.execute("INSERT INTO foo(t) VALUES (?)", &[&ts])?;

            let from: OffsetDateTime = db.query_row("SELECT t FROM foo", [], |r| r.get(0))?;

            db.execute("DELETE FROM foo", [])?;

            assert_eq!(from, ts);
        }
        Ok(())
    }

    #[test]
    fn test_sqlite_functions() -> Result<()> {
        let db = checked_memory_handle()?;
        let result: Result<OffsetDateTime> =
            db.query_row("SELECT CURRENT_TIMESTAMP", [], |r| r.get(0));
        assert!(result.is_ok());
        Ok(())
    }
}
