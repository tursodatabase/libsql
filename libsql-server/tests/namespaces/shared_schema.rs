use hyper::StatusCode;
use insta::{assert_debug_snapshot, assert_json_snapshot};
use libsql::Database;
use serde_json::json;
use tempfile::tempdir;
use tokio::time::Duration;
use turmoil::Builder;

use crate::common::{http::Client, net::TurmoilConnector};

use super::make_primary;

macro_rules! assert_all_eq {
    ($first:expr, $( $rest:expr ),+ $(,)?) => {
        $(
            assert_eq!($first, $rest);
        )+
    };
}

async fn get_schema_version(ns: &str) -> i64 {
    let db = Database::open_remote_with_connector(
        format!("http://{ns}.primary:8080"),
        String::new(),
        TurmoilConnector,
    )
    .unwrap();
    let conn = db.connect().unwrap();
    let mut rows = conn.query("PRAGMA schema_version", ()).await.unwrap();
    rows.next().await.unwrap().unwrap().get::<i64>(0).unwrap()
}

async fn check_schema(ns: &str) -> Vec<String> {
    let db = Database::open_remote_with_connector(
        format!("http://{ns}.primary:8080"),
        String::new(),
        TurmoilConnector,
    )
    .unwrap();
    let conn = db.connect().unwrap();
    let mut rows = conn.query("SELECT * from sqlite_schema", ()).await.unwrap();
    let mut out = Vec::new();
    while let Some(row) = rows.next().await.unwrap() {
        out.push(format!("{row:?}"));
    }

    out
}

async fn check_data(ns: &str) -> Vec<String> {
    let db = Database::open_remote_with_connector(
        format!("http://{ns}.primary:8080"),
        String::new(),
        TurmoilConnector,
    )
    .unwrap();
    let conn = db.connect().unwrap();
    let mut rows = conn.query("SELECT * from test", ()).await.unwrap();
    let mut out = Vec::new();
    while let Some(row) = rows.next().await.unwrap() {
        out.push(format!("{row:?}"));
    }
    out
}

async fn http_get(url: &str) -> String {
    let client = Client::new();
    client.get(url).await.unwrap().body_string().await.unwrap()
}

#[test]
fn perform_schema_migration() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(100000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async {
        let client = Client::new();
        client
            .post(
                "http://primary:9090/v1/namespaces/schema/create",
                json!({"shared_schema": true }),
            )
            .await
            .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/ns1/create",
                json!({"shared_schema_name": "schema" }),) .await .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/ns2/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();

        let schema_db = Database::open_remote_with_connector(
            "http://schema.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap();
        let schema_version_before = get_schema_version("schema").await;
        let schema_conn = schema_db.connect().unwrap();
        schema_conn
            .execute("create table test (c)", ())
            .await
            .unwrap();

        while get_schema_version("schema").await == schema_version_before {
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        // all schemas are the same
        assert_debug_snapshot!(check_schema("ns1").await);
        assert_debug_snapshot!(check_schema("ns2").await);
        assert_debug_snapshot!(check_schema("schema").await);

        // check all schema versions are same as primary schema db
        let expected_schema_version = 1;
        assert_all_eq!(expected_schema_version, get_schema_version("schema").await, get_schema_version("ns1").await, get_schema_version("ns2").await);

        let resp = http_get("http://primary:9090/v1/namespaces/schema/migrations").await;
        assert_eq!(
            resp,
            r#"{"schema_version":4,"migrations":[{"job_id":1,"status":"RunSuccess"}]}"#
        );

        let resp = http_get("http://primary:9090/v1/namespaces/schema/migrations/1").await;
        assert_eq!(resp, r#"{"job_id":1,"status":"RunSuccess","progress":[{"namespace":"ns1","status":"RunSuccess","error":null},{"namespace":"ns2","status":"RunSuccess","error":null}]}"#);

        // we add a new namespace and expect it to have the same schema version and schema
        client
            .post(
                "http://primary:9090/v1/namespaces/ns3/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();
        assert_debug_snapshot!(check_schema("ns3").await);
        assert_eq!(expected_schema_version, get_schema_version("ns3").await);

        // we will perform a new migration and expect the new and old databases to have same schema
        schema_conn
            .execute("create table test2 (c)", ())
            .await
            .unwrap();

        while get_schema_version("schema").await == schema_version_before {
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        // all schemas are the same
        assert_debug_snapshot!(check_schema("ns1").await);
        assert_debug_snapshot!(check_schema("ns2").await);
        assert_debug_snapshot!(check_schema("ns3").await);
        assert_debug_snapshot!(check_schema("schema").await);

        // check all schema versions are same as primary schema db
        let expected_schema_version = 2;
        assert_all_eq!(expected_schema_version, get_schema_version("schema").await, get_schema_version("ns1").await, get_schema_version("ns2").await, get_schema_version("ns3").await);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn no_job_created_when_migration_job_is_invalid() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(100000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async {
        let client = Client::new();
        client
            .post(
                "http://primary:9090/v1/namespaces/schema/create",
                json!({"shared_schema": true }),
            )
            .await
            .unwrap();

        let schema_db = Database::open_remote_with_connector(
            "http://schema.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap();
        let schema_conn = schema_db.connect().unwrap();
        assert_debug_snapshot!(schema_conn
            .execute_batch("create table test (c); create table test (c)")
            .await
            .unwrap_err());

        let resp = client
            .get("http://primary:9090/v1/namespaces/schema/migrations/1")
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::NOT_FOUND);
        assert_debug_snapshot!(resp.json_value().await.unwrap());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn migration_contains_txn_statements() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(100000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async {
        let client = Client::new();
        client
            .post(
                "http://primary:9090/v1/namespaces/schema/create",
                json!({"shared_schema": true }),
            )
            .await
            .unwrap();

        let schema_db = Database::open_remote_with_connector(
            "http://schema.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap();
        let schema_conn = schema_db.connect().unwrap();
        assert_debug_snapshot!(schema_conn
            .execute_batch("begin; create table test (c)")
            .await
            .unwrap_err());

        let resp = client
            .get("http://primary:9090/v1/namespaces/schema/migrations/1")
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::NOT_FOUND);
        assert_debug_snapshot!(resp.json_value().await.unwrap());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn dry_run_failure() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(100000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async {
        let client = Client::new();
        client
            .post(
                "http://primary:9090/v1/namespaces/schema/create",
                json!({"shared_schema": true }),
            )
            .await
            .unwrap();

        client
            .post(
                "http://primary:9090/v1/namespaces/ns1/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/ns2/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();

        let schema_db = Database::open_remote_with_connector(
            "http://schema.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap();
        let schema_conn = schema_db.connect().unwrap();
        let schema_version_before = get_schema_version("schema").await;
        schema_conn
            .execute_batch("create table test (c)")
            .await
            .unwrap();

        while get_schema_version("schema").await == schema_version_before {
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        let ns1_db = Database::open_remote_with_connector(
            "http://ns1.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap();
        let ns1_conn = ns1_db.connect().unwrap();
        ns1_conn
            .execute("insert into test values (NULL)", ())
            .await
            .unwrap();

        // we are creating a new table with a constraint on the column. when the dry run is
        // executed against the schema or ns2, it works, because test is empty there, but it should
        // fail on ns1, because it test contains a row with NULL.
        schema_conn
            .execute_batch("create table test2 (c NOT NULL); insert into test2 select * from test")
            .await
            .unwrap();

        loop {
            let resp = client
                .get("http://primary:9090/v1/namespaces/schema/migrations/2")
                .await
                .unwrap()
                .json_value()
                .await
                .unwrap();
            if resp["status"].as_str().unwrap() == "RunFailure" {
                break;
            }
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        let resp = client
            .get("http://primary:9090/v1/namespaces/schema/migrations/2")
            .await
            .unwrap()
            .json_value()
            .await
            .unwrap();
        assert_json_snapshot!(resp);

        // all schemas are the same (test2 doesn't exist)
        assert_debug_snapshot!(check_schema("ns1").await);
        assert_debug_snapshot!(check_schema("ns2").await);
        assert_debug_snapshot!(check_schema("schema").await);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn perform_data_migration() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(100000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async {
        let client = Client::new();
        client
            .post(
                "http://primary:9090/v1/namespaces/schema/create",
                json!({"shared_schema": true }),
            )
            .await
            .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/ns1/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/ns2/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();

        let schema_db = Database::open_remote_with_connector(
            "http://schema.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap();
        let schema_version_before = get_schema_version("schema").await;
        let schema_conn = schema_db.connect().unwrap();
        schema_conn
            .execute("create table test (c)", ())
            .await
            .unwrap();
        schema_conn
            .execute("insert into test values (42)", ())
            .await
            .unwrap();

        while get_schema_version("schema").await == schema_version_before {
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        let get_row = |ns: String| async move {
            let db = Database::open_remote_with_connector(
                format!("http://{ns}.primary:8080"),
                String::new(),
                TurmoilConnector,
            )
            .unwrap();
            let conn = db.connect().unwrap();
            let mut rows = conn.query("SELECT * from test", ()).await.unwrap();
            rows.next().await.unwrap().unwrap().get::<i64>(0).unwrap()
        };

        assert_all_eq!(
            42,
            get_row("ns1".to_string()).await,
            get_row("ns2".to_string()).await,
            get_row("schema".to_string()).await
        );

        // new db added to the linked schema also should have the data
        client
            .post(
                "http://primary:9090/v1/namespaces/ns3/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();
        assert_eq!(42, get_row("ns3".to_string()).await);

        // new insertions in the schema db should be propagated to all
        schema_conn
            .execute("insert into test values (43)", ())
            .await
            .unwrap();
        assert_debug_snapshot!(check_data("ns1").await);
        assert_debug_snapshot!(check_data("ns2").await);
        assert_debug_snapshot!(check_data("ns3").await);
        assert_debug_snapshot!(check_data("schema").await);

        // updates and deletes should be propagated
        schema_conn
            .execute("update test set c = 50 where c = 42", ())
            .await
            .unwrap();
        schema_conn
            .execute("delete from test where c = 43", ())
            .await
            .unwrap();

        let get_row = |ns: String| async move {
            let db = Database::open_remote_with_connector(
                format!("http://{ns}.primary:8080"),
                String::new(),
                TurmoilConnector,
            )
            .unwrap();
            let conn = db.connect().unwrap();
            let mut rows = conn.query("SELECT * from test", ()).await.unwrap();
            rows.next().await.unwrap().unwrap().get::<i64>(0).unwrap()
        };

        assert_all_eq!(
            50,
            get_row("ns1".to_string()).await,
            get_row("ns2".to_string()).await,
            get_row("ns3".to_string()).await,
            get_row("schema".to_string()).await
        );

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn conflicting_data_migration() {
    let mut sim = Builder::new()
        .simulation_duration(Duration::from_secs(100000))
        .build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async {
        let client = Client::new();
        client
            .post(
                "http://primary:9090/v1/namespaces/schema/create",
                json!({"shared_schema": true }),
            )
            .await
            .unwrap();
        client
            .post(
                "http://primary:9090/v1/namespaces/ns1/create",
                json!({"shared_schema_name": "schema" }),
            )
            .await
            .unwrap();

        let schema_conn = Database::open_remote_with_connector(
            "http://schema.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap()
        .connect()
        .unwrap();
        let schema_version_before = get_schema_version("schema").await;
        schema_conn
            .execute("create table test (c)", ())
            .await
            .unwrap();
        schema_conn
            .execute("insert into test values (42)", ())
            .await
            .unwrap();

        while get_schema_version("schema").await == schema_version_before {
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        assert_debug_snapshot!(check_data("ns1").await);
        assert_debug_snapshot!(check_data("schema").await);

        let ns1_conn = Database::open_remote_with_connector(
            "http://ns1.primary:8080",
            String::new(),
            TurmoilConnector,
        )
        .unwrap()
        .connect()
        .unwrap();

        // insert some data and try to create a conflicting index
        ns1_conn
            .execute("insert into test values (42)", ())
            .await
            .unwrap();

        // create unique index on the column
        schema_conn
            .execute("create unique index idx on test (c)", ())
            .await
            .unwrap();

        loop {
            let resp = client
                .get("http://primary:9090/v1/namespaces/schema/migrations/3")
                .await
                .unwrap()
                .json_value()
                .await
                .unwrap();
            if resp["status"].as_str().unwrap() == "RunFailure" {
                break;
            }
            tokio::time::sleep(Duration::from_millis(100)).await;
        }

        let resp = client
            .get("http://primary:9090/v1/namespaces/schema/migrations/3")
            .await
            .unwrap()
            .json_value()
            .await
            .unwrap();
        assert_json_snapshot!(resp);

        // query if ns1 has zero index
        let mut rows = ns1_conn
            .query("select count(*) from sqlite_schema where type='index'", ())
            .await
            .unwrap();
        let c = rows.next().await.unwrap().unwrap().get::<i64>(0).unwrap();
        assert_eq!(c, 0);

        Ok(())
    });

    sim.run().unwrap();
}
