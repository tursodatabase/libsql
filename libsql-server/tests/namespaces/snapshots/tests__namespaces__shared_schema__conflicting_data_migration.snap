---
source: libsql-server/tests/namespaces/shared_schema.rs
expression: "check_data(\"ns1\").await"
---
[
    "Row { cols: [Col { name: Some(\"c\"), decltype: None }], inner: [Integer { value: 42 }] }",
]
